### DISCUSSION TITLE
Changing the default branch

### DISCUSSION THREAD

**Discussion by chas-iot** [sqm-autorate/sqm-autorate#117] [General]
Which branch should be tagged as the default branch in github?

The `main` branch is very old and no work takes place in that branch.

The development work takes place in feature branches of `develop/main` and releases are 'published' in `testing/lua-threads`.

Initially, I thought that we should tag `testing/lua-threads`, now I suggest `develop/main` as we do point users to install from the testing branch and the installation script does default to download from that branch too.

**Comment by chas-iot** [comment]
I also propose that we rename testing/lua-threads to testing

I think that the process should be
1. branch testing from testing/lua-threads
2. update the scripts and docs in develop/main to point to testing 
3. cherry pick the above commit into testing/lua-threads
4. fast forward testing to testing/lua-threads
5. Announce on OpenWRT
6. (after a day or two) delete the testing/lua-threads branch in the repo

**Reply by chas-iot** [reply]
hmmm! the issue is that the branch names contain `/`
this means that all branches have to have the same number of `/` so that the documentation relative links work

so either
1. rename testing/lua-threads to testing/main, or
2. rename all branches to change `/` to `-` eg. develop/main -> develop-main (or a similar separator)

**Reply by richb-hanover** [reply]
I agree with this plan. Other thoughts:

-  I don't much care about the naming of the branches, but long term, I worry that the presence of "/" characters will bite us and force the need to rush out a silly patch. From that standpoint, I would be happier just having names use hyphens. (And we should just bite the bullet now.)

- A wacky (non-)constraint. Could we find a name for the branch that we want people to look at (current placeholder is `testing/lua-threads`) so that it's listed _before_ the many, many develop/... branches? As it is, the dropdown of branches shows:

main
archive/...
archive/...
develop/...
develop/...
...infinity :-)...
develop/...
testing/lua-threads...

So it's not "good enough" to call it testing/aardvark. (Obviously, a bunch of the develop/... branches will soon go away, so maybe this isn't such a big deal.) Thanks.

**Reply by Fail-Safe** [reply]
IMHO, testing/lua-threads was always supposed to be promoted to "main" eventually. It's just another one of those things that got lost in the shuffle of the rate at which we were coding. Once we agree that testing/lua-threads is stable enough, we should promote it to "main" and let that be the public stable branch it was always supposed to be.

**Comment by richb-hanover** [comment]
@chas-iot , @Fail-Safe @dlakelan @Lochnair  I see all the good work Charles is doing to coalesce his new work into `develop/main`. I wonder if we should screw up our courage and simply rename that branch to `main`. Reasons:

- It has been our long-term plan
- The README on `main` then would describe our current best thinking
- It provides a stable platform to publicize
- It gives us permission to deprecate (or eliminate) the other branches that aren't receiving as much love

It doesn't have to happen immediately, but we should talk about whether this is a good step. Thanks

**Comment by chas-iot** [comment]
@richb-hanover unfortunately there is a lot more to changing the default branch than just the name.

Our branch names contain '`/`', eg.  `develop/main` and `testing/lua-threads`. This has an impact on all of the relative references in our documentation. A lot of the references will break if one branch does not have `/` while the others do

**Reply by richb-hanover** [reply]
Oh, yes. I'm quite aware of the work this will require. But I'm advocating that we start the plan...

**Comment by Fail-Safe** [comment]
A `main` branch has been created containing 'stable' code.
